Connect to dotnet-profiler to collect testwise coverage

r2h2 opened the pull request.

r2h2 committed:
Add .NET options for zeroMQ in extension config

r2h2 committed:
Differentiate between technoliges when constructing Teamscale calls

r2h2 committed:
Re-use teamscale URL in config for dotnet profiler
Omit previous profiler configuration, splitting request construction for .net and sap as .net needs to send a body, too

r2h2 committed:
Improvements from ILP on dotnet profiler communications

r2h2 committed:
Merge branch 'main' into zeroMQ-options

r2h2 committed:
Different labels for TS and web server

r2h2 committed:
Log errors on console and user popup on connection errors

r2h2 committed:
update action is supported for .NET

r2h2 committed:
show technology options also on load

rottkit committed:
fix technology on load

rottkit commented on background.js:
```diff
@@ -43,13 +43,17 @@ chrome.runtime.onMessage.addListener(
 
-		if (request.data.request === internalRequestKinds.resetUser) {
-			queryTeamscale(tsTiaApiActions.reset);
+		if(configOptions[technologyId] === technologies.sap ){
```
Seems the formatting is not in line with the used one (inconsistent spacing before and after brackets)

r2h2 replied:
Thanks, I applied the default JavaScript formatter within IntelliJ Ultimate to all JS files, equivalently for HTML files.

rottkit commented on background.js:
```diff
@@ -179,7 +189,20 @@ function testRunUpdateCallListener(details, tabId, tab) {
 
-	queryTeamscale(action, details.tabId, testNameWithParameter, testOutcomeToTeamscaleTestExecutionResult(updatedIterationActionResult.outcome));
+	if(configOptions[technologyId] === technologies.dotnet ){
+		// HACK
```
I would assume this comment a little more explaining. Is it nonetheless working (i.e., stopping and resuming a test)?
Or will coverage after having pressed "pause" be lost -- if so, add a log message so that the user knows?

Please refer this TODO/HACK-comment to a Jira ticket if this should be fixed in the future. If no fix is to be done in the future it is rather no HACK but the behaviour should be documented?

r2h2 replied:
You're absolutely right. I reformulated this code to log a warning explaining the user that the current implementation stops coverage recording when tests are paused in ADO. Resuming is not supported, at least at the moment (and this is fine with ILP).

rottkit commented on background.js:
```diff
@@ -179,7 +189,20 @@ function testRunUpdateCallListener(details, tabId, tab) {
+		queryProfiler(action, details.tabId, testNameWithParameter, testOutcomeToTeamscaleTestExecutionResult(updatedIterationActionResult.outcome));
+	} else if (configOptions[technologyId] === technologies.sap ){
+		queryTeamscale(action, details.tabId, testNameWithParameter, testOutcomeToTeamscaleTestExecutionResult(updatedIterationActionResult.outcome));
+	} else {
```
This is a little inconsistent with the other locations where new distinctions by cases were introduced.
Here is the only location where we catch "invalid" technologies. Can it happen that the method is called with an invalid technology?

rottkit replied:
ah.. ok in `resolveUserNameOfTesterAndTriggerRecordingStart` we also check for an invalid technology

r2h2 replied:
From my point of view, these are the two entry points where the technology could be invalid. In fact, this is a "should never happen" case, but might happen if we add more technologies and forget to implement them in some of the distinctions.

rottkit commented on background.js:
```diff
@@ -225,17 +248,41 @@ function getParameterDefinitionsPerIteration(updateRequest) {
+
+	request.onreadystatechange = () => handleResponse(request, action);
+	request.onerror = () => handleError(request);
+	var result = {"result":status};
```
Could be `const`?

r2h2 replied:
thanks, done :)

rottkit commented on background.js:
```diff
@@ -225,17 +248,41 @@ function getParameterDefinitionsPerIteration(updateRequest) {
 }
 
+function queryProfiler(action, tabId, extendedName, status, sapTestKey) {
+	if (action !== tsTiaApiActions.reset && action !== tsTiaApiActions.log && (!tabId || !testCaseIdByTab[tabId])) {
```
Extract this `if` and the one starting in line `263` as they clone the conditions ins `queryTeamscale`?

r2h2 replied:
Thanks, see new function:
`function validateTestId(action, tabId) {
    if (action !== tsTiaApiActions.reset && action !== tsTiaApiActions.log && (!tabId || !testCaseIdByTab[tabId])) {
        throw 'Could not obtain testId from tabId "' + tabId + '".';
    }
}`

rottkit replied:
Thanks, did you decide to not extract the latter part (after checking the configuration entries) on purpose or did you just missed that one?

r2h2 replied:
Indeed, I missed the second if. I added it in the extracted function, renamed the function and adapted the calling functions accordingly

rottkit commented on background.js:
```diff
@@ -225,17 +248,41 @@ function getParameterDefinitionsPerIteration(updateRequest) {
 
+function queryProfiler(action, tabId, extendedName, status, sapTestKey) {
+	if (action !== tsTiaApiActions.reset && action !== tsTiaApiActions.log && (!tabId || !testCaseIdByTab[tabId])) {
+		throw 'Could not obtain testId from tabId "' + tabId + '".';
```
Do you consider these `throw`s only for debugging?
If yes, everything 👍. If not, should we make them visible to the user (in the extensions "popup")?

rottkit replied:
Please resolve, if this is as intended.

r2h2 replied:
Yes, I consider these `throws` only for debugging purposes. -> Resolve

rottkit commented on background.js:
```diff
@@ -463,8 +569,9 @@ function isTestRunnerApiCall(details) {
 function setDefaultOptions() {
 	const standardValues = {};
-	standardValues[tsServerOptionId] = 'https://teamscale.example.org/';
+	standardValues[serverOptionId] = 'https://teamscale.example.org/';
```
With this change also the data structure of the objected stored in the chrome store is changed (see line 580).
Old installations will not work out-of-the-box anymore. However, this might not be that grave as it is not actively used?

But the semantic of the attribute is different depending on the used technology. Not sure if we should not just introduce a separate one.

rottkit replied:
Can you please comment on this as well?

r2h2 replied:
I introduced a new attribute for the web server which communicates with the profiler and renamed this property back to its original name.

rottkit commented on background.js:
```diff
@@ -225,17 +248,41 @@ function getParameterDefinitionsPerIteration(updateRequest) {
 	}
 
-	if (!configOptions[tsServerOptionId] || !configOptions[tsProjectOptionId] || !configOptions[sapUserOptionId]) {
+	if (!configOptions[serverOptionId] || !configOptions[tsProjectOptionId] || !configOptions[sapUserOptionId]) {
```
I am not entirely sure, whether we should reuse the existing Teamscale-URL option. (see also comment in `background.js` at the bottom)

r2h2 replied:
See above 👍

rottkit commented on background.js:
```diff
@@ -246,10 +293,46 @@ function queryTeamscale(action, tabId, extendedName, status, sapTestKey) {
+	let additionalParameter = '';
+	if (action === tsTiaApiActions.update) {
+		additionalParameter = '&result=' + status + "&extended-name=" + encodeURI(extendedName);
+	}
```
`additionalParameter` is reported unused

rottkit replied:
Only by my WebStrom and not from Teamscale: https://build.cqse.eu/teamscale/delta.html#findings/ts-ados-test-listener-extension/?from=zeroMQ-options%3A1656268598261&to=main%3A1656268598261&showMergeFindings=true

rottkit replied:
I think you missed a few comments @r2h2

r2h2 replied:
I appended this to the URL.

rottkit commented on background.js:
```diff
@@ -246,10 +293,46 @@ function queryTeamscale(action, tabId, extendedName, status, sapTestKey) {
+		additionalParameter = '&result=' + status + "&extended-name=" + encodeURI(extendedName);
+	}
+
+	const teamscaleUrl = assertStringEndsWith(configOptions[serverOptionId], '/');
```
Naming wrong due to c&p left-over?

rottkit replied:
I think you missed a few comments @r2h2 ;-)

r2h2 replied:
Indeed, thanks

rottkit commented on background.js:
```diff
@@ -246,10 +293,46 @@ function queryTeamscale(action, tabId, extendedName, status, sapTestKey) {
+
+	const teamscaleUrl = assertStringEndsWith(configOptions[serverOptionId], '/');
+
+	const testOutput = 'Follow this link to view test run in Azure DevOps:\n' + adosTestRunUrlByTab[tabId];
```
unused?

rottkit replied:
I think you missed a few comments @r2h2

r2h2 replied:
Resolved :-)

rottkit commented on background.js:
```diff
@@ -246,10 +293,46 @@ function queryTeamscale(action, tabId, extendedName, status, sapTestKey) {
+	} else if (action === tsTiaApiActions.log) {
+		throw tsTiaApiActions.log + ' action not supported, yet, for .NET.'
+	} else if (action === tsTiaApiActions.stop) {
+		// using "end" legacy end point to support the Java proiler use-case, too.
```
Unclear comment, please elaborate for others who have no in-depth knowledge.
Which Java-Profiler is meant? What would be the alternative to the legacy one?

rottkit replied:
I think you missed a few comments @r2h2

r2h2 replied:
Thanks, added more details to the comment

rottkit commented on background.js:
```diff
@@ -246,10 +293,46 @@ function queryTeamscale(action, tabId, extendedName, status, sapTestKey) {
+	const serviceUrl = teamscaleUrl + 'test/';
+
+	if (action === tsTiaApiActions.reset) {
+		throw tsTiaApiActions.reset + ' action not supported, yet, for .NET.'
```
Minor: Interesting word ordering in this sentence ;-)

r2h2 replied:
Got rid of two commas :-)

rottkit commented on background.js:
```diff
@@ -296,6 +380,21 @@ function handleTeamscaleResponse(request, action) {
 }
 
+function handleError(request){
+	const actionString = '🛑 Connection Refused';
```
I guess `Refused` is not always true. There might be other `error`s?
https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/error_event#live_example

r2h2 replied:
True, I suggest a more abstract formulation

rottkit commented on background.js:
```diff
@@ -296,6 +380,21 @@ function handleTeamscaleResponse(request, action) {
 
+function handleError(request){
+	const actionString = '🛑 Connection Refused';
+	const text = 'Connecting to ' + configOptions[serverOptionId] + 'failed. Is the server running?';
```
Is the second sentence also a good tipp for the `queryTeamscale` from which this also could be evoked?

r2h2 replied:
I intentionally formulated this generic enough so that it matches also for the Teamscale server case. Theoretically, it is possible that Teamscale instances are not running. WDYT? 
To make it a bit more clear, I rephrased the second sentence as follows: "Is the server running and reachable?"

rottkit replied:
👍

Please add the missing space before "failed"

r2h2 replied:
Sure, done 👍

rottkit commented on background.js:
```diff
@@ -296,6 +380,21 @@ function handleTeamscaleResponse(request, action) {
+
+	persistAndBroadcastSingleEvent(event);
+
+	throw 'Connecting to ' + configOptions[serverOptionId] + 'failed. Is the server running?'
```
Is that `throw` needed? What happens, if we leave that out?

r2h2 replied:
I wanted to have this on the DEV console, too (mainly for debugging purposes). As the logging works well, I removed this `throw`.

rottkit commented on options.js:
```diff
@@ -18,4 +19,29 @@ function saveChanges(inputElement) {
 }
 
+function showTechnologyOptions(){
+    var technology = document.getElementById("technology");
+    var sapInput = document.getElementById("sap-input");
+    var tsProject = document.getElementById("ts-project-input");
+    var tsServerDescription = document.getElementById("ts-server-description");
+    var webServerDescription = document.getElementById("web-server-description");
```
use `const`s instead of `var`s?

rottkit replied:
Change naming to an `element` suffix?

r2h2 replied:
Used `const`s and added `Element` as suffixes.

rottkit commented on options.js:
```diff
@@ -18,4 +19,29 @@ function saveChanges(inputElement) {
+    var webServerDescription = document.getElementById("web-server-description");
+
+    if ( technology.value == 'dotnet')
+      {
```
Please apply a formatter to not have the inconsistencies in how brackets are set. :-) I guess, earlier we used how our WebStorms were configured in the past.

r2h2 replied:
See comment above on `background.js`

rottkit commented on options.js:
```diff
@@ -18,4 +19,29 @@ function saveChanges(inputElement) {
+     }
+}
+
+document.getElementById("technology").addEventListener("change", showTechnologyOptions);
```
Can we use the element's variable/`const` from above instead of repeating `document.getElementById("technology")`?

r2h2 replied:
Yes, I moved the declarations to global space, to be able to reuse the `const`

rottkit reviewed (commented):
Thanks for this work! 🙏 🤩 
Finished my first round of review. Will probably not be able to answer (at least on computer) during the upcoming week. Unfortunately I could not test it manually on the environment we talked about on Friday, since logging in is currently not possible for me (some technical error).

r2h2 committed:
Implement Jakob's feedback

r2h2 commented:
@rottkit Thank you for your first round of review! I addressed your comments in the code and replied to the comments in this merge request.
Regarding testing on SAP I agree that we need to accomplish this before publishing a new version. If I find some time tomorrow, I will have a look.

r2h2 committed:
Add IntelliJ formatting file

rottkit commented:
Thanks for your rework @r2h2 
I think you missed a bunch of the comments. Please have a look again :-)

r2h2 committed:
Include server option ID check in validateTestOptionId function

r2h2 committed:
Include server option ID check in validateTestOptionId function

r2h2 committed:
Merge remote-tracking branch 'origin/zeroMQ-options' into zeroMQ-options

# Conflicts:
#	background.js

r2h2 committed:
Separate Teamscale and Profiler URL fields

r2h2 committed:
fix casing of sap technology option
Lower-case is necessary as this is the value given to the config option

r2h2 commented:
Thanks @rottkit for your helpful feedback and sorry for missing several comments in the first round!

I think I have addressed all your comments.

I tested the current version at ILP (fine) and MR. Unfortunately, the SCOV main switch is currently disabled at F57 and I reached out to their pool to let it turn on again. At least, the test start and stop events were correctly detected and I also got feedback from Teamscale that (a) the main switch is not enabled and (b) there was no coverage recording session for my user which could have been stopped (which is expected in case of disabled SCOV). I will re-test when the SAP team has re-enabled the SCOV and keep you posted.

r2h2 committed:
Generalize popup (does not need to be SAP specific)

r2h2 commented:
@rottkit Thanks again for having a look at my changes! Now this is ready for review round 2 :)

Update from MR:
I tested this implementation on a SAP testing environment and all events (start, stop, update) and requests seem to work as expected, i.e., coverage recording start + stop, error on multiple active test sessions in ADO, report upload to Teamscale.
The test scenario was a bit limited as I have no permission to execute any monitored code on the system, so the coverage report was empty. Still, I think we can expect the event and data processing pipeline to work. WDYT?

r2h2 committed:
need ? as separator

r2h2 committed:
Do not re-set testCaseID
If task is created directly during test execution, the previous logic failed to forward  further events

r2h2 committed:
For .NET, send test case "ID - TITLE" to profiler to show TITLE in Teamscale
This is a temporary workaround until we've got a better implementation in the profiler
